plugins/qcheck-stm/src/stm_of_ir.ml

```diff
@@ -43,13 +43,294 @@ let subst inst ty =
   in
   aux ty
 
+let lazy_force =
+  let open Gospel in
+  let open Tterm_helper in
+  let vs_name = Ident.create ~loc:Location.none "Lazy.force"
+  and vs_ty = Ttypes.fresh_ty_var "a" in
+  let lazy_force = mk_term (Tvar { vs_name; vs_ty }) None Location.none in
+  fun t ->
+    Tterm_helper.(
+      mk_term (Tapp (Symbols.fs_apply, [ lazy_force; t ])) None Location.none)
+
+let ocaml_of_term cfg t =
+  let open Ortac_core.Ocaml_of_gospel in
+  let open Reserr in
+  try term ~context:cfg.Cfg.context t |> ok with W.Error e -> error e
+
+let subst_term ~gos_t ?(old_lz = false) ~old_t ?(new_lz = false) ~new_t term =
+  let exception ImpossibleSubst of (Gospel.Tterm.term * [ `New | `Old ]) in
+  let rec aux cur_lz cur_t term =
+    let open Gospel.Tterm in
+    let next = aux cur_lz cur_t in
+    match term.t_node with
+    | Tconst _ -> term
+    | Tvar { vs_name; vs_ty } when vs_name = gos_t -> (
+        match cur_t with
+        | Some cur_t ->
+            let t = { term with t_node = Tvar { vs_name = cur_t; vs_ty } } in
+            if cur_lz then lazy_force t else t
+        | None ->
+            raise (ImpossibleSubst (term, if cur_t = new_t then `New else `Old))
+        )
+    | Tvar _ -> term
+    | Tapp (ls, terms) -> { term with t_node = Tapp (ls, List.map next terms) }
+    | Tfield (t, ls) -> { term with t_node = Tfield (next t, ls) }
+    | Tif (cnd, thn, els) ->
+        { term with t_node = Tif (next cnd, next thn, next els) }
+    | Tlet (vs, t1, t2) -> { term with t_node = Tlet (vs, next t1, next t2) }
+    | Tcase (t, brchs) ->
+        {
+          term with
+          t_node =
+            Tcase
+              ( next t,
+                List.map
+                  (fun (p, ot, t) -> (p, Option.map next ot, next t))
+                  brchs );
+        }
+    | Tquant (q, vs, t) -> { term with t_node = Tquant (q, vs, next t) }
+    | Tbinop (o, l, r) -> { term with t_node = Tbinop (o, next l, next r) }
+    | Tnot t -> { term with t_node = Tnot (next t) }
+    | Told t -> aux old_lz old_t t
+    | Ttrue -> term
+    | Tfalse -> term
+  in
+  let open Reserr in
+  try ok (aux new_lz new_t term)
+  with ImpossibleSubst (t, b) ->
+    error
+      ( Impossible_term_substitution
+          (Fmt.str "%a" Gospel.Tterm_printer.print_term t, b),
+        t.t_loc )
+
+let str_of_ident = Fmt.str "%a" Gospel.Identifier.Ident.pp
+
+let mk_cmd_pattern value =
+  let pat_args = function
+    | None -> punit
+    | Some x -> ppat_var (noloc (str_of_ident x))
+  in
+  let args =
+    match value.args with
+    | [] -> None
+    | [ x ] -> Some (pat_args x)
+    | xs -> List.map pat_args xs |> ppat_tuple |> Option.some
+  in
+  let name = String.capitalize_ascii (str_of_ident value.id) |> lident in
+  ppat_construct name args
+
+let next_state_case config state_ident value =
```

Comment:
Should we document somewhere around that the `idx` part of the returned value is _not_ sorted as it is in the end keeping the order of the `modifies` rather than the `ensures`?